[21797] Add Fast DDS v3 Migration Guide (#929)
eProsima/Fast-DDS-docs: docs/notes/migration_guide.rst

.. include:: ../03-exports/aliases.include
.. include:: ../03-exports/aliases-api.include
.. include:: ../03-exports/roles.include

> Migration Guide from Fast DDS v2 to Fast DDS v3

Reviewer: Title too long for this section, I would keep it shorter
```suggestion
Migration Guide to Fast DDS v3
```

===============================================

This document aims to help during the migration process from eProsima *Fast DDS version* 2 to *Fast DDS version* 3.

> For more information about all the updates, please refer to the [release notes](https://fast-dds.docs.eprosima.com/en/latest/notes/notes.html).

Reviewer: I would add a reference in supported_versions.rst as `versions_v3` to tag it here
```suggestion
For more information about all the updates, please refer to the :ref:`release notes <release_notes_v3>`.
```

Migration Steps
---------------

The following steps describe the possible changes that your project may require to migrate to *Fast DDS v3.0.0*:

- :ref:`step-1-update-the-package-name-and-cmake-configuration`
- :ref:`step-2-update-dependencies`
- :ref:`step-3-ensure-compability-with-related-products`
- :ref:`step-4-apply-namespace-changes`
- :ref:`step-5-migrate-public-headers`
- :ref:`step-6-handle-removed-or-private-headers`
- :ref:`step-7-update-api-methods`
- :ref:`step-8-update-structs-enums-and-variables`
- :ref:`step-9-refactor-examples`

.. _step-1-update-the-package-name-and-cmake-configuration:

> Step 1: Update the Package Name and CMake Configuration

Reviewer: ```suggestion
Step 1: Update the package name and CMake configuration
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> 1. CMake Project Name: Rename the CMake project from ``fastrtps`` to ``fastdds``.

Reviewer: ```suggestion
1. CMake project name: Rename the CMake project from ``fastrtps`` to ``fastdds``.
```

> 2. Environment Variables:

Reviewer: ```suggestion
2. Environment variables:
```

   * Rename ``FASTRTPS_DEFAULT_PROFILES_FILE`` to ``FASTDDS_DEFAULT_PROFILES_FILE``.
   * The configuration file for loading profiles is now ``DEFAULT_FASTDDS_PROFILES.xml``.

> 3. Update CMake File Names on Windows:

Reviewer: ```suggestion
3. Update CMakeFile names on Windows:
```

   * Rename ``fastrtps.manifest.in`` to ``fastdds.manifest.in``.
   * Rename ``fastrtps-config.cmake`` to ``fastdds-config.cmake``.
   * Rename ``fastrtps.rc`` to ``fastdds.rc``.

.. _step-2-update-dependencies:

> Step 2: Update Dependencies

Reviewer: ```suggestion
Step 2: Update dependencies
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^

Fast DDS v3 is only compatible with Fast CDR v2.
If you are not using Fast CDR as :ref:`third-party <third-party_libraries-options>`, please ensure that your local dependencies are up-to-date.
Refer to the :ref:`library deprendencies table <dependencies_compatibilities_build_system_dependencies>` to verify version compatibility for all Fast DDS library dependencies.

.. _step-3-ensure-compability-with-related-products:

> Step 3: Ensure Compability with Related Products

Reviewer: ```suggestion
Step 3: Ensure compatibility with related products
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Fast DDS v3 requires Fast DDS Gen v4 for code generation. Make sure to regenerate types using this compatible version.

> For other compatibility requirements with related products (e.g., Shapes Demo, Discovery Server), refer to the table of [products compatibility](https://fast-dds.docs.eprosima.com/en/latest/notes/versions.html#eprosima-products-compatibility), which outlines version compatibility across the Fast DDS ecosystem.

Reviewer: ```suggestion
For other compatibility requirements with related products (e.g., Shapes Demo, Discovery Server), refer to the table of :ref:`products compatibility <dependencies_compatibilities_product_compatibility>`, which outlines version compatibility across the Fast DDS ecosystem.
```

Verify these versions and update accordingly to avoid any integration issues.

.. _step-4-apply-namespace-changes:

> Step 4: Apply Namespace Changes

Reviewer: ```suggestion
Step 4: Apply namespace changes
```

Reviewer: Not applied

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> 1. Namespace Migration:

Reviewer: ```suggestion
1. Namespace migration:
```

   * Update all ``eprosima::fastrtps::`` namespace references to ``eprosima::fastdds::``.
   * Move built-in topics ``SubscriptionBuiltinTopicData``, ``PublicationBuiltinTopicData``, and ``ParticipantBuiltinTopicData`` from ``fastdds::dds::builtin::`` to ``fastdds::dds::``.
   * Move ``Duration_t`` and ``c_TimeInfinite`` references to ``dds::``.
   * Move ``Time_t.hpp`` references from ``eprosima::fastdds::`` to ``eprosima::fastdds::dds``.

   Ensure you update these namespace references across your code to avoid compilation errors.

> 2. Renamed Types:

Reviewer: ```suggestion
2. Renamed types:
```

   * Change ``EventKindBits::`` references to ``EventKind::``.
   * Change ``EventKindEntityId::`` references to ``EntityId::``.
   * Change ``StatisticsEventKind::`` references to ``statistics::EventKind::``.

   Refactor the type references as outlined above to maintain compatibility with the new version.

.. _step-5-migrate-public-headers:

> Step 5: Migrate Public Headers

Reviewer: ```suggestion
Step 5: Migrate public headers
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> 1. File Location headers:

Reviewer: ```suggestion
1. Header files location:
```

   All the headers in ``include/fastrtps`` were migrated to ``include/fastdds``.
   In particular, the following list includes headers that have been relocated to different paths or whose implementations have been incorporated into other headers:

   .. list-table::
      :header-rows: 1

>       * - Fast DDS 2.x file *include* path
>         - Fast DDS v3.0.0 file *include* path

Reviewer: ```suggestion
      * - Fast DDS v2 file *include* path
        - Fast DDS v3 file *include* path
```

        - ``fastdds/dds/core/status/SampleRejectedStatus.hpp``
      * - ``fastrtps/participant/Participant.h``
        - ``fastdds/rtps/participant/RTPSParticipant.hpp``
      * - ``fastrtps/transport/TCPv4TransportDescriptor.h``
        - ``fastdds/rtps/transport/TCPv4TransportDescriptor.hpp``
      * - ``fastrtps/transport/TCPv6TransportDescriptor.h``
        - ``fastdds/rtps/transport/TCPv6TransportDescriptor.hpp``
      * - ``fastrtps/transport/UDPv4TransportDescriptor.h``
        - ``fastdds/rtps/transport/UDPv4TransportDescriptor.hpp``
      * - ``fastrtps/transport/UDPv6TransportDescriptor.h``
        - ``fastdds/rtps/transport/UDPv6TransportDescriptor.hpp``
      * - ``fastrtps/transport/UDPTransportDescritpor.h``
        - ``fastdds/rtps/transport/UDPTransportDescritpor.hpp``
      * - ``fastrtps/transport/TCPTransportDescritpor.h``
        - ``fastdds/rtps/transport/TCPTransportDescritpor.hpp``
      * - ``fastdds/rtps/common/Time_t.hpp in namespace{fastdds}``
        - ``fastdds/dds/core/Time_t.hpp in namespace{fastdds::dds}``

   Also, the ``fixed_size_string.hpp`` implementation has been migrated from Fast DDS package to Fast CDR.

> 2. File Extensions:

Reviewer: ```suggestion
2. File extensions:
```

   Rename file extensions from `.h` to `.hpp`.

.. _step-6-handle-removed-or-private-headers:

> Step 6: Handle Removed or Private Headers

Reviewer: ```suggestion
Step 6: Handle removed or private headers
```

* PublisherAttributes.hpp
* SubscriberAttributes.hpp
* ProxyPool.hpp
* Semaphore.hpp
* MessageReceiver.hpp
* BuiltinProtocols.hpp
* shared_mutex.hpp
* StringMatching.hpp
* TimeConversion.hpp
* DBQueue.hpp
* ResourceEvent.hpp
* TimedEvent.hpp
* WriterProxyData.hpp
* ReaderProxyData.hpp
* ParticipantProxyData.hpp
* XML Parser API
* UnitsParser.hpp
* RTPSMessageGroup.hpp
* RTPSMessageCreator.hpp
* CDRMessage.hpp
* StatefulPersistentReader.hpp
* StatefulReader.hpp
* StatelessPersistentReader.hpp
* StatelessReader.hpp
* PersistentWriter.hpp
* StatefulPersistentWriter.hpp
* StatefulWriter.hpp
* StatelessPersistentWriter.hpp
* StatelessWriter.hpp
* logging.h
* Exception.h
* Cryptography.h
* Authentication.h
* AccessControl.h
* SecurityException.h
* ChangeForReader.hpp
* ReaderLocator.hpp
* ReaderProxy.hpp
* ServerAttributes.hpp
* TopicAttributes.hpp
* TypeLookupService.hpp

If your project previously included any of these headers, you will need to modify your implementation.
Since these headers are now private, you should replace their usage with public alternatives or refactor the related code to ensure it does not depend on private headers.

.. _step-7-update-api-methods:

> Step 7: Update API Methods

Reviewer: ```suggestion
Step 7: Update API methods
```

^^^^^^^^^^^^^^^^^^^^^^^^^^

The table below contains the list of API changes, showing the previous methods and the corresponding new ones introduced in Fast DDS v3.
The new API methods achieve the same functionality, even though the signature of the method is different from the deprecated one.

.. list-table::
   :header-rows: 1

   * - Deprecated methods
     - New methods
   * - ``xmlparser::XMLProfileManager::library_settings(LibrarySettingsAttributes&)``

>      - ``DomainParticipantFactory::get_instance()->set_library_settings(const LibrarySettings&)``

Reviewer: Ideally, we should reference the API reference in the new methods, but it is a lot of work...

Author: But I wouldn't be able to reference the deleted methods, right?

Reviewer: Exactly

Reviewer 2: This will be done in a follow-up PR. See [21797] (https://eprosima.easyredmine.com/issues/21797)

   * - ``TypeSupport::calculate_serialized_size_provider(const void* const, DataRepresentationId_t)``
     - ``TypeSupport::calculate_serialized_size(const void* const, DataRepresentationId_t)``
   * - ``get_key(void, InstanceHandle_t*, bool)``
     - ``compute_key(SerializedPayload_t&, InstanceHandle_t&, bool)``
   * - ``DynamicPubSubType::createData()``
     - ``DynamicPubSubType::create_data()``
   * - ``DynamicPubSubType::deleteData(void*)``
     - ``DynamicPubSubType::delete_data(void*)``
   * - ``DynamicPubSubType::getKey(const void* const, InstanceHand*, bool)``
     - ``DynamicPubSubType::compute_key(const void* const, InstanceHand&, bool)``
   * - ``DynamicPubSubType::getSerializedSizeProvider(const void* const, DataRepresentationId_t)``
     - ``DynamicPubSubType::calculate_serialized_size(const void* const, DataRepresentationId_t)``

Review your code for any APIs marked with the ``FASTDDS_DEPRECATED`` and ``FASTDDS_TODO_BEFORE`` macros.
Note that these deprecated APIs have been removed in Fast DDS v3.
Make the necessary updates to your implementation to ensure compatibility with the new version.

.. _step-8-update-structs-enums-and-variables:

> Step 8: Update Structs, Enums, and Variables

Reviewer: ```suggestion
Step 8: Update structs, enums, and variables
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

As part of the Fast DDS migration, several structs, enums, and variables have been updated. You will need to modify your code to reflect these changes:

1. Enum and Variable Changes:

>     * ``DiscoveryProtocol_t`` is now ``DiscoveryProtocol``.

Reviewer: ```suggestion
    * Rename ``DiscoveryProtocol_t`` to ``DiscoveryProtocol``.
```

    * Rename ``heartbeatResponseDelay`` to ``heartbeat_response_delay``.
    * Rename ``initialAcknackDelay`` to ``initial_acknack_delay``.
    * Rename ``expectsInlineQos`` to ``expects_inline_qos``.
    * Rename ``m_typeSize`` to ``max_serialized_type_size``.
    * Rename ``m_isGetKeyDefined`` to ``is_compute_key_provided``.
    * Rename ``m_topicDataTypeName`` to ``topic_data_typename``.

2. Extend Built-in Topics:

    * ``SubscriptionBuiltinTopicData`` has been extended with additional fields to mimic those of ``ReaderProxyData``.
    * ``PublicationBuiltinTopicData`` has been extended with additional fields to mimic those of ``WriterProxyData``.
    * ``ParticipantBuiltinTopicData`` has been extended to include the product version and fields from ``ParticipantProxyData``.

3. Other Struct Changes:

    * ``SendBuffersAllocationAttributes`` has a new attribute to define the allocation configuration of the ``NetworkBuffers``.
    * ``TypeConsistencyQos`` has been removed from ``DataReader``, and the ``TypeConsistencyEnforcementQosPolicy`` and ``DataRepresentationQosPolicy`` have been added.

.. _step-9-refactor-examples:

> Step 9: Refactor Examples

Reviewer: ```suggestion
Step 9: Examples refactor
```

^^^^^^^^^^^^^^^^^^^^^^^^^

All examples have been refactored to follow a consistent structure across the Fast DDS project.
This includes renaming files, restructuring classes, and updating the overall format.
Additionally, it is important to note that some examples have been removed, renamed, or had significant changes to their options and configurations.
If you have integrated any of these examples into your own implementation, carefully review the updated examples to ensure compatibility with your project.

All the examples have been refactored to follow the same structure:

* File names, guards, and classes follow new format.
* Detailed and well-formed README.md with example explanation.

> * Example structured in applications, stopped by `SIGTERM` signal.

Reviewer: ```suggestion
* Example structured in applications, stopped by ``SIGTERM`` signal.
```

> Hello World
> """""""""""

Reviewer: What do you think about linking the examples here? Apply in the remain examples if considered

```suggestion
`Hello World <https://github.com/eProsima/Fast-DDS/tree/master/examples/cpp/hello_world>`_
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
```

For more information about all the updates, please refer to the :ref:`release notes <release_notes>`.

Migration Steps
---------------

The following steps describe the possible changes that your project may require to migrate to *Fast DDS v3.0.0*:

- :ref:`step-1-update-the-package-name-and-cmake-configuration`
- :ref:`step-2-update-dependencies`
- :ref:`step-3-ensure-compatibility-with-related-products`
- :ref:`step-4-apply-namespace-changes`
- :ref:`step-5-migrate-public-headers`
- :ref:`step-6-handle-removed-or-private-headers`
- :ref:`step-7-update-api-methods`
- :ref:`step-8-update-structs-enums-and-variables`
- :ref:`step-9-refactor-examples`

.. _step-1-update-the-package-name-and-cmake-configuration:

Step 1: Update the package name and CMake configuration
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1. CMake project name: Rename the CMake project from ``fastrtps`` to ``fastdds``.
2. Environment variables:

   * Rename ``FASTRTPS_DEFAULT_PROFILES_FILE`` to ``FASTDDS_DEFAULT_PROFILES_FILE``.
   * The configuration file for loading profiles is now ``DEFAULT_FASTDDS_PROFILES.xml``.

> 3. Update CMakeFile names on Windows:

Reviewer: I messed up previous revision here, sorry
```suggestion
3. Update CMake file names on Windows:
```

    through the filter.
  - For the Default filter, they represent the maximum and minimum value message indexes that are read.

`Custom Payload Pool <https://github.com/eProsima/Fast-DDS/tree/master/examples/cpp/custom_payload_pool>`_
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

Refactor the CustomPayloadPoolExample example with the current new example format.

`Delivery Mechanism <https://github.com/eProsima/Fast-DDS/tree/master/examples/cpp/delivery_mechanisms>`_
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

In this delivery mechanisms example, the key changes are:

* Loans and data-sharing compatible: bounded types, final extensibility.
* Loans mechanism for data management.
* Option to select all delivery mechanisms.

`Discovery Server <https://github.com/eProsima/Fast-DDS/tree/master/examples/cpp/discovery_server>`_
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

Refactor the DiscoveryServerExample example with the current new example format.

> `Flow Controller <https://github.com/eProsima/Fast-DDS/tree/master/examples/cpp/flow_controller>`_
> """"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

Reviewer: ```suggestion
`Flow Controller <https://github.com/eProsima/Fast-DDS/tree/master/examples/cpp/flow_control>`_
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
```
